tinygrad/tensor.py

```diff
@@ -1317,6 +1317,29 @@ def unflatten(self, dim:int, sizes:Tuple[int,...]):
     dim = self._resolve_dim(dim)
     return self.reshape(self.shape[:dim] + sizes + self.shape[dim+1:])
 
+  def roll(self, shifts:Union[int, Tuple[int, ...]], dims:Union[int, Tuple[int, ...]]) -> Tensor:
+    """
+    Roll the tensor along specified dimension(s).
+    The rolling operation is circular, meaning that elements that go beyond the edge are wrapped around to the beginning of the dimension.
+
+    ```python exec="true" source="above" session="tensor" result="python"
+    print(Tensor.rand(3, 4, 1).roll(shifts=1, dims=0))
+    ```
+    ```python exec="true" source="above" session="tensor" result="python"
+    print(Tensor.rand(3, 4, 1).roll(shifts=-1, dims=0))
+    ```
+    """
+    dims, shifts = (dims, ) if isinstance(dims, int) else dims, (shifts, ) if isinstance(shifts, int) else shifts
+    dims = tuple(i % len(self.shape) for i in dims)
+    all_dims = range(len(self.shape))
+    all_shifts = [shifts[dims.index(i)] % self.shape[i] if i in dims else 0 for i in all_dims]
+    rolled = self
+    for shift, dim in zip(all_shifts, all_dims):
```

Comment:
enumerate

Reply:
Reduced the line count a lot, still using range len for the shape though. I can replace it with enumerate but then I'd just do:
`for i, _ in enumerate()`
for which I don't really see point in doing